### PR TITLE
expression: avoid using datum in `HashCode()`.

### PR DESCRIPTION
**Pull request by winoros**
Simplify the logic of `HashCode`.
```
Constant: ConstantFlag+encoded value
Column: ColumnFlag+encoded value
ScalarFunction: SFFlag+encoded function name + encoded arg_1 + encoded arg_2 + ... 
```
PTAL @zz-jason @shenli @lamxTyler

### CODE REVIEW

**Comment by shenli** [comment]
Please fix the CI.

File: expression/expression.go [hunk 1]
```diff
@@ -28,6 +28,12 @@ import (
 	"github.com/pingcap/tidb/types/json"
 )
 
+const (
```
Comment (zz-jason) [inline expression/expression.go]: add some comments about the usage of theses flags

**Comment by winoros** [comment]
PTAL @shenli @zz-jason

**Pull request merged**